viskores/cont/testing/UnitTestSplineEvaluate.cxx

```diff
@@ -0,0 +1,329 @@
+      }
+      else if (i == 2)
+      {
+        //quadratic clustering near 1.
+        xcoords = fillCoords(d[0], [](viskores::FloatDefault t) { return 1 - (1 - t) * (1 - t); });
+        ycoords = fillCoords(d[1], [](viskores::FloatDefault t) { return 1 - (1 - t) * (1 - t); });
+        zcoords = fillCoords(d[2], [](viskores::FloatDefault t) { return 1 - (1 - t) * (1 - t); });
+      }
+      else if (i == 3)
+      {
+        //exponential
+        xcoords = fillCoordsExp(d[0]);
+        ycoords = fillCoordsExp(d[1]);
+        zcoords = fillCoordsExp(d[2]);
+      }
+
+      auto ds = builder.Create(xcoords, ycoords, zcoords);
+      viskores::cont::Invoker invoker;
+      viskores::cont::ArrayHandle<viskores::FloatDefault> fieldArray;
+      invoker(GenerateData{}, ds.GetCoordinateSystem(), fieldArray);
+      ds.AddPointField("field", fieldArray);
+      dataSets.push_back(ds);
+    }
+  }
+
+  return dataSets;
+}
+
+viskores::cont::ArrayHandle<viskores::FloatDefault> CreateLinearInterpolationResult(
+  const viskores::cont::DataSet& ds,
+  const viskores::cont::ArrayHandle<viskores::Vec3f>& points)
+{
+  viskores::cont::DataSetBuilderExplicit builder;
+  std::vector<viskores::Id> ids(points.GetNumberOfValues());
+  std::iota(ids.begin(), ids.end(), 0);
+  auto idsArray = viskores::cont::make_ArrayHandle(ids, viskores::CopyFlag::On);
+
+  auto ptDataSet = builder.Create(points, viskores::CellShapeTagVertex(), 1, idsArray);
+
+  viskores::filter::resampling::Probe probe;
+  probe.SetActiveField("field");
+  probe.SetGeometry(ptDataSet);
+  auto output = probe.Execute(ds);
+
+  viskores::cont::ArrayHandle<viskores::FloatDefault> result;
+  ds.GetField("field").GetData().AsArrayHandle(result);
+  return result;
+}
+
+template <typename SplineEvalType>
+void CompareToLinear(SplineEvalType& splineEval,
+                     const viskores::cont::ArrayHandle<viskores::Vec3f>& points,
+                     const viskores::cont::ArrayHandle<viskores::FloatDefault>& expectedValues,
+                     const viskores::cont::DataSet& ds)
+{
+  viskores::cont::Invoker invoke;
+  viskores::cont::ArrayHandle<viskores::FloatDefault> results;
+
+  //Compare against linear interpolation.
+  auto linearResult = CreateLinearInterpolationResult(ds, points);
+
+  EvalWorklet evalWorklet;
+  invoke(evalWorklet, points, splineEval, results);
+
+  auto splinePortal = results.ReadPortal();
+  auto linearPortal = linearResult.ReadPortal();
+  auto expectedPortal = expectedValues.ReadPortal();
+  for (viskores::Id i = 0; i < splinePortal.GetNumberOfValues(); i++)
+  {
+    auto truth = expectedPortal.Get(i);
+    auto linearValue = linearPortal.Get(i);
+    auto splineValue = splinePortal.Get(i);
+
+    auto diffLinear = viskores::Abs(truth - linearValue);
+    auto diffSpline = viskores::Abs(truth - splineValue);
+    auto factor = diffLinear / diffSpline;
+
+    // spline interpolation should be better.
+    VISKORES_TEST_ASSERT(factor > 1.0f,
```

Comment:
Why not just directly compare the values?

```suggestion
    // spline interpolation should be better.
    VISKORES_TEST_ASSERT(diffSpline <= diffLinear,
```

Wouldn't that work better if you hit a sample very close to a grid point and both are very close to 0?